Content.Server/_Goobstation/Possession/PossessionSystem.cs

```diff
@@ -0,0 +1,279 @@
+using Content.Server.Stunnable;
+using Content.Shared.Administration.Logs;
+using Content.Shared.CombatMode.Pacification;
+using Content.Shared.Coordinates;
+using Content.Shared.Database;
+using Content.Shared.Examine;
+using Content.Shared.Ghost;
+using Content.Shared.Mind;
+using Content.Shared.Mindshield.Components;
+using Content.Shared.Mobs.Systems;
+using Content.Shared.Popups;
+using Content.Shared.Tag;
+using Content.Shared.Zombies;
+using Robust.Server.Containers;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Containers;
+using Robust.Shared.Map;
+using Robust.Shared.Spawners;
+using Robust.Shared.Timing;
+using Content.Server.Bible.Components;
+using Robust.Shared.Prototypes;
+using Content.Shared._Goobstation.Religion;
+
+namespace Content.Server._Goobstation.Possession;
+
+public sealed partial class PossessionSystem : EntitySystem
+{
+    [Dependency] private readonly IGameTiming _timing = default!;
+    [Dependency] private readonly MobStateSystem _mobState = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedMindSystem _mind = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly StunSystem _stun = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly ContainerSystem _container = default!;
+    [Dependency] private readonly ISharedAdminLogManager _admin = default!;
+    [Dependency] private readonly ActionsSystem _action = default!;
+    [Dependency] private readonly PolymorphSystem _polymorph = default!;
+    [Dependency] private readonly TagSystem _tag = default!;
+
+    private ProtoId<TagPrototype> _cannotSuicideAnyTag = "CannotSuicideAny"; // DeltaV - Don't use literals.
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<PossessedComponent, MapInitEvent>(OnInit);
+        SubscribeLocalEvent<PossessedComponent, ComponentRemove>(OnComponentRemoved);
+
+        SubscribeLocalEvent<PossessedComponent, ExaminedEvent>(OnExamined);
+
+        SubscribeLocalEvent<PossessedComponent, EndPossessionEarlyEvent>(OnEarlyEnd);
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        var query = EntityQueryEnumerator<PossessedComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if (_timing.CurTime >= comp.PossessionEndTime)
+                RemComp<PossessedComponent>(uid);
+
+            comp.PossessionTimeRemaining = comp.PossessionEndTime - _timing.CurTime;
+        }
+    }
+
+    private void OnInit(Entity<PossessedComponent> possessed, ref MapInitEvent args)
+    {
+        if (!HasComp<WeakToHolyComponent>(possessed))
+            AddComp<WeakToHolyComponent>(possessed).AlwaysTakeHoly = true;
+        else
+            possessed.Comp.WasWeakToHoly = true;
```

Comment:
`if (!HasComp) AddComp` is just `EnsureComp` with extra steps

Reply:
Actually not in this case, because we use different behaviour when one pre-exists